token/services/identity/role/role.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/identity"
+	idriver "github.com/hyperledger-labs/fabric-token-sdk/token/services/identity/driver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/logging"
+	"github.com/pkg/errors"
+	"go.uber.org/zap/zapcore"
+)
+
+type localMembership interface {
+	DefaultNetworkIdentity() driver.Identity
+	IsMe(id driver.Identity) bool
+	GetIdentityInfo(label string, auditInfo []byte) (idriver.IdentityInfo, error)
+	GetIdentifier(id driver.Identity) (string, error)
+	GetDefaultIdentifier() string
+	RegisterIdentity(config driver.IdentityConfiguration) error
+	IDs() ([]string, error)
+}
+
+// Role models a role whose identities are anonymous
+type Role struct {
+	logger          logging.Logger
+	roleID          identity.RoleType
+	networkID       string
+	localMembership localMembership
+	nodeIdentity    driver.Identity
+}
+
+func NewRole(logger logging.Logger, roleID identity.RoleType, networkID string, nodeIdentity driver.Identity, localMembership localMembership) *Role {
+	return &Role{
+		logger:          logger,
+		roleID:          roleID,
+		networkID:       networkID,
+		localMembership: localMembership,
+		nodeIdentity:    nodeIdentity,
+	}
+}
+
+func (r *Role) ID() identity.RoleType {
+	return r.roleID
+}
+
+// GetIdentityInfo returns the identity information for the given identity identifier
+func (r *Role) GetIdentityInfo(id string) (idriver.IdentityInfo, error) {
+	if r.logger.IsEnabledFor(zapcore.DebugLevel) {
+		r.logger.Debugf("[%s] getting info for [%s]", r.networkID, id)
+	}
+
+	info, err := r.localMembership.GetIdentityInfo(id, nil)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "[%s] failed to get identity for [%s]", r.networkID, id)
+	}
+	return info, nil
+}
+
+// RegisterIdentity registers the given identity
+func (r *Role) RegisterIdentity(config driver.IdentityConfiguration) error {
+	return r.localMembership.RegisterIdentity(config)
+}
+
+func (r *Role) IdentityIDs() ([]string, error) {
+	return r.localMembership.IDs()
+}
+
+// MapToIdentity returns the identity for the given argument
+func (r *Role) MapToIdentity(v driver.WalletLookupID) (driver.Identity, string, error) {
+	switch vv := v.(type) {
+	case driver.Identity:
+		return r.mapIdentityToID(vv)
+	case []byte:
+		return r.mapIdentityToID(vv)
+	case string:
+		return r.mapStringToID(vv)
+	default:
+		return nil, "", errors.Errorf("identifier not recognised, expected []byte or driver.Identity, got [%T], [%s]", v, string(debug.Stack()))
+	}
+}
+
+func (r *Role) mapStringToID(v string) (driver.Identity, string, error) {
+	defaultNetworkIdentity := r.localMembership.DefaultNetworkIdentity()
+	defaultIdentifier := r.localMembership.GetDefaultIdentifier()
+
+	if r.logger.IsEnabledFor(zapcore.DebugLevel) {
+		r.logger.Debugf("[%s] mapping string identifier for [%s,%s], default identities [%s:%s]",
+			r.networkID,
+			v,
+			hash.Hashable(v).String(),
```

Comment:
`.String()` will be called if needed lazily. Don't call it already.

Reply:
okay